azure/services/roleassignments/roleassignments.go

```diff
@@ -32,54 +34,65 @@ import (
 )
 
 const azureBuiltInContributorID = "b24988ac-6180-42a0-ab88-20f7382dd24c"
+const serviceName = "roleassignments"
 
 // RoleAssignmentScope defines the scope interface for a role assignment service.
 type RoleAssignmentScope interface {
 	azure.ClusterDescriber
-	RoleAssignmentSpecs() []azure.RoleAssignmentSpec
+	azure.AsyncStatusUpdater
+	RoleAssignmentSpecs() []azure.ResourceSpecGetter
 }
 
 // Service provides operations on Azure resources.
 type Service struct {
 	Scope RoleAssignmentScope
 	client
+	async.Reconciler
 	virtualMachinesClient        virtualmachines.Client
 	virtualMachineScaleSetClient scalesets.Client
 }
 
 // New creates a new service.
 func New(scope RoleAssignmentScope) *Service {
+	client := newClient(scope)
 	return &Service{
 		Scope:                        scope,
-		client:                       newClient(scope),
+		client:                       client,
 		virtualMachinesClient:        virtualmachines.NewClient(scope),
 		virtualMachineScaleSetClient: scalesets.NewClient(scope),
+		Reconciler:                   async.New(scope, client, client),
 	}
 }
 
 // Reconcile creates a role assignment.
 func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "roleassignments.Service.Reconcile")
 	defer done()
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
 	for _, roleSpec := range s.Scope.RoleAssignmentSpecs() {
-		switch roleSpec.ResourceType {
+		rs := roleSpec.(*RoleAssignmentSpec)
+		scope := fmt.Sprintf("/subscriptions/%s/", s.Scope.SubscriptionID())
+		rs.Scope = scope
+		contributorRoleDefinitionID := fmt.Sprintf("/subscriptions/%s/providers/Microsoft.Authorization/roleDefinitions/%s", s.Scope.SubscriptionID(), azureBuiltInContributorID)
+		rs.RoleDefinitionID = contributorRoleDefinitionID
+		switch rs.ResourceType {
```

Comment:
We could create a getter function such as `RoleAssignmentResourceType(name)` and add it to the role assignment scope to determine if you need to reconcile a VM or a VMSS.